crates/bitwarden-crypto/src/safe/data_envelope.rs

```diff
@@ -0,0 +1,656 @@
+        data: T,
+        wrapping_key: &Ids::Symmetric,
+        ctx: &mut crate::store::KeyStoreContext<Ids>,
+    ) -> Result<(Self, EncString), DataEnvelopeError>
+    where
+        T: Serialize + SealableVersionedData,
+    {
+        let (envelope, cek) = Self::seal(data, ctx)?;
+
+        let wrapped_cek = ctx
+            .wrap_symmetric_key(*wrapping_key, cek)
+            .map_err(|_| DataEnvelopeError::EncryptionError)?;
+
+        Ok((envelope, wrapped_cek))
+    }
+
+    /// Seals a struct into an encrypted blob, and returns the encrypted blob and the
+    /// content-encryption-key.
+    fn seal_ref<T>(
+        data: &T,
+        namespace: &DataEnvelopeNamespace,
+    ) -> Result<(DataEnvelope, XChaCha20Poly1305Key), DataEnvelopeError>
+    where
+        T: Serialize + SealableVersionedData,
+    {
+        let mut cek = XChaCha20Poly1305Key::make();
+
+        // Serialize the message
+        let serialized_message =
+            SerializedMessage::encode(&data).map_err(|_| DataEnvelopeError::EncodingError)?;
+        ensure_equal!(serialized_message.content_type(), coset::iana::CoapContentFormat::Cbor => DataEnvelopeError::UnsupportedContentFormat);
+        let serialized_and_padded_message = pad_cbor(serialized_message.as_bytes())
+            .map_err(|_| DataEnvelopeError::EncodingError)?;
+
+        // Build the COSE headers
+        let mut protected_header = coset::HeaderBuilder::new()
+            .key_id(cek.key_id.to_vec())
+            .content_type(CONTENT_TYPE_PADDED_CBOR.to_string())
+            .value(
+                DATA_ENVELOPE_NAMESPACE,
+                ciborium::Value::Integer(Integer::from(namespace.as_i64())),
+            )
+            .build();
+        protected_header.alg = Some(coset::Algorithm::PrivateUse(XCHACHA20_POLY1305));
+
+        // Encrypt the message
+        let mut nonce = [0u8; xchacha20::NONCE_SIZE];
+        let encrypt0 = coset::CoseEncrypt0Builder::new()
+            .protected(protected_header)
+            .create_ciphertext(&serialized_and_padded_message, &[], |data, aad| {
+                let ciphertext =
+                    crate::xchacha20::encrypt_xchacha20_poly1305(&(*cek.enc_key).into(), data, aad);
+                nonce = ciphertext.nonce();
+                ciphertext.encrypted_bytes().to_vec()
+            })
+            .unprotected(coset::HeaderBuilder::new().iv(nonce.to_vec()).build())
+            .build();
+
+        // Serialize the COSE message
+        let envelope_data = encrypt0
+            .to_vec()
+            .map(CoseEncrypt0Bytes::from)
+            .map_err(|_| DataEnvelopeError::EncodingError)?;
+
+        // Disable key operations other than decrypt on the CEK
+        cek.disable_key_operation(coset::iana::KeyOperation::Encrypt)
+            .disable_key_operation(coset::iana::KeyOperation::WrapKey)
+            .disable_key_operation(coset::iana::KeyOperation::UnwrapKey);
+
+        Ok((DataEnvelope { envelope_data }, cek.clone()))
```

Comment:
```suggestion
        Ok((DataEnvelope { envelope_data }, cek))
```

I don't think the clone is needed?